src/org/sosy_lab/java_smt/solvers/z3/Z3UserPropagator.java

```diff
@@ -0,0 +1,185 @@
+  @Override
+  protected Z3UserPropagator freshWrapper(long lctx) {
+    return this;
+  }
+
+  // TODO: This method is called if the user registers a function (currently not
+  //  possible) and the solver instantiates the registered function: if the solver
+  //  instantiates "forall x: f(x)" at x=y, then f(y) will get created.
+  @Override
+  protected void createdWrapper(long le) {
+  }
+
+  @Override
+  protected void decideWrapper(long lvar, int bit, boolean is_pos) {
+    // We currently only allow tracking of decision on boolean formulas
+    // so we ignore the <bit> parameter
+    userPropagator.onDecision(encapsulate(lvar), is_pos);
+  }
+
+
+  // ===========================================================================
+  // Function calls from JavaSMT's side (mostly calls to the smt backend)
+  // ===========================================================================
+
+  @Override
+  public void registerExpression(BooleanFormula exprToWatch) {
+    Native.propagateAdd(this, ctx, solver, javainfo, creator.extractInfo(exprToWatch));
+  }
+
+  @Override
+  public void notifyOnKnownValue() {
+    registerFixed();
+  }
+
+  @Override
+  public void notifyOnDecision() {
+    registerDecide();
+  }
+
+  @Override
+  public void notifyOnFinalCheck() {
+    registerFinal();
+  }
+
+  @Override
+  public void propagateConflict(BooleanFormula[] conflictExpressions) {
+    propagateConsequence(conflictExpressions, manager.getBooleanFormulaManager().makeFalse());
+  }
+
+  @Override
+  public void propagateConsequence(BooleanFormula[] assignedLiterals, BooleanFormula consequence) {
+    long[] emptyEqs = new long[0];
+    Native.propagateConsequence(this, ctx, solver, javainfo,
+        assignedLiterals.length,
+        extractInfoFromArray(assignedLiterals),
+        emptyEqs.length,
+        emptyEqs,
+        emptyEqs,
+        creator.extractInfo(consequence)
+    );
+  }
+
+  private enum Z3LBool {
+    FALSE(-1),
+    UNDEFINED(0),
+    TRUE(1);
+
+    final int value;
+
+    Z3LBool(int value) {
+      this.value = value;
+    }
+  }
+
+  @Override
+  public boolean propagateNextDecision(BooleanFormula expr, Optional<Boolean> value) {
+    Z3LBool phase = value.map(pBoolean -> (pBoolean ? Z3LBool.TRUE : Z3LBool.FALSE))
+        .orElse(Z3LBool.UNDEFINED);
+    int index = 0; // Only relevant for bitvector expressions, which are not supported yet.
+    return Native.propagateNextSplit(this, ctx, solver, javainfo, creator.extractInfo(expr),
+        index, phase.value);
+  }
+
+  private long[] extractInfoFromArray(BooleanFormula[] formulaArray) {
+    long[] formulaInfos = new long[formulaArray.length];
+    for (int i = 0; i < formulaArray.length; i++) {
+      if (formulaArray[i] != null) {
+        formulaInfos[i] = creator.extractInfo(formulaArray[i]);
+      }
+    }
+    return formulaInfos;
+  }
+
+  private BooleanFormula encapsulate(long z3Expr) {
+    // We fill the lowest 3 bits that are always 0 due pointer alignment.
```

Comment:
This is an interesting observation. It might not be universally true, but valid for most systems. Lets keep it. :-)

Reply:
Do you support 32-bit versions of Z3? If so, this code could break. I just added this little trick because the standard values result in a lot of collisions, especially on smaller maps.

Reply:
A user can provide its own version of Z3, including 32bit.
Are you sure about collisions here? Java should already use some bitwise and/or modulo operator internally in the hashmap.

Reply:
When profiling, I just noticed that the hashmap code for tree nodes was run all the time and tree nodes are only created if there are a lot of collisions (linear buckets get replaced by tree buckets if the bucket gets too full).
The hashcode for long is computed as `(l >> 32) ^ l` but the upper range tends to be constant so that the lower 3 bits end up being constant as well. The hashmap itself uses a power of two module so it retains only the lower bits of the hashcode.
As this would be really bad on its own, the hashmap tries to do some bit-mixing on the hashcode before truncating it down, but apparently it does not work too well here. That being said, my profiling was restricted to a single internal benchmark and the Nqueens examples, so I don't have extensive data.

A simple solution is to just shift by 2 rather than 3 to also make it correct on 32 bit systems (assuming pointers are well-aligned).

Reply:
Thanks for the nice explanation.

A sound way could be a full bitwise rotation when computing a hashable value: `(l << 61) | (l >> 3)`.
That should work well on your observation and is independent of a specific architecture. :-)

Reply:
I didn't think about bitwise rotation... a much nicer solution indeed :).